### PR TITLE
upstream: degraded stats and hc event logs

### PR DESCRIPTION
**Pull request by snowp**
Adds a basic gauge for number of degraded members as well as health check event logs for
when a host is degraded/no longer degraded

*Risk Level*: Low
*Testing*: UTs
*Docs Changes*: Documented new stat + proto docs
*Release Notes*: n/a
Part of #5063

### CODE REVIEW

File: docs/root/configuration/cluster_manager/cluster_stats.rst [hunk 1]
```diff
@@ -78,6 +78,7 @@ Every cluster has a statistics tree rooted at *cluster.<name>.* with the followi
   upstream_flow_control_drained_total, Counter, Total number of times the upstream connection drained and resumed reads from downstream
   membership_change, Counter, Total cluster membership changes
   membership_healthy, Gauge, Current cluster healthy total (inclusive of both health checking and outlier detection)
+  membership_healthy, Gauge, Current cluster degraded total
```
Comment (junr03) [inline docs/root/configuration/cluster_manager/cluster_stats.rst]:
```suggestion
  membership_degraded, Gauge, Current cluster degraded total
```

File: include/envoy/upstream/health_checker.h [hunk 2]
```diff
@@ -99,6 +99,21 @@ class HealthCheckEventLogger {
...
+   * @param host supplies the host that generated the event.
+   */
+  virtual void logDegraded(envoy::data::core::v2alpha::HealthCheckerType health_checker_type,
+                           const HostDescriptionConstSharedPtr& host) PURE;
+  /**
+   * Log a no degraded healthy host event.
+   * @param health_checker_type supplies the type of health checker that generated the event.
+   * @param host supplies the host that generated the event.
+   */
+  virtual void logNoMoreDegraded(envoy::data::core::v2alpha::HealthCheckerType health_checker_type,
```
Comment (junr03) [inline include/envoy/upstream/health_checker.h]:
```suggestion
  virtual void logNoLongerDegraded(envoy::data::core::v2alpha::HealthCheckerType health_checker_type,
```

Seems a bit less clunky?

File: api/envoy/data/core/v2alpha/health_check_event.proto [hunk 3]
```diff
@@ -30,6 +30,12 @@ message HealthCheckEvent {
 
     // Host failure.
     HealthCheckFailure health_check_failure_event = 7;
+
+    // Healthy host became degraded.
+    DegradedHealthyHost degraded_healthy_host = 8;
+
+    // A degraded host returned to being healthy.
+    NotDegradedHealthyHost not_degraded_healthy_host = 9;
```
Comment (junr03) [inline api/envoy/data/core/v2alpha/health_check_event.proto]: I suggest naming this NoLongerDegraded to match the function. But up to you!

File: source/common/upstream/health_checker_base_impl.cc [hunk 4]
```diff
@@ -353,5 +378,38 @@ void HealthCheckEventLoggerImpl::logAddHealthy(
   file_->write(fmt::format("{}\n", json));
 }
 
+void HealthCheckEventLoggerImpl::logDegraded(
+    envoy::data::core::v2alpha::HealthCheckerType health_checker_type,
+    const HostDescriptionConstSharedPtr& host) {
+  envoy::data::core::v2alpha::HealthCheckEvent event;
```
Comment (junr03) [inline source/common/upstream/health_checker_base_impl.cc]: could we take all of the logging functions common attribute sets and move them to a common function? Similar to what @nezdolik is doing in #5517 with `setCommonEventParams`

File: test/common/upstream/upstream_impl_test.cc [hunk 5]
```diff
@@ -1264,39 +1265,66 @@ TEST(StaticClusterImplTest, HealthyStat) {
...
+
+  // Mark the endpoint as unhealthy. This should decrement the degraded stat.
+  cluster.prioritySet().hostSetsPerPriority()[0]->hosts()[1]->healthFlagSet(
+      Host::HealthFlag::FAILED_ACTIVE_HC);
+  health_checker->runCallbacks(cluster.prioritySet().hostSetsPerPriority()[0]->hosts()[1],
+                               HealthTransition::Changed);
+  EXPECT_EQ(0UL, cluster.prioritySet().hostSetsPerPriority()[0]->healthyHosts().size());
+  EXPECT_EQ(0UL, cluster.prioritySet().hostSetsPerPriority()[0]->degradedHosts().size());
+  EXPECT_EQ(0UL, cluster.info()->stats().membership_healthy_.value());
+  EXPECT_EQ(0UL, cluster.info()->stats().membership_degraded_.value());
```
Comment (junr03) [inline test/common/upstream/upstream_impl_test.cc]: should we also cover the case of going from degraded to healthy?

File: test/common/upstream/health_checker_impl_test.cc [hunk 6]
```diff
@@ -591,6 +591,7 @@ TEST_F(HttpHealthCheckerImplTest, Degraded) {
   // We start off as healthy, and should go degraded after receiving the degraded health response.
   EXPECT_CALL(*test_sessions_[0]->interval_timer_, enableTimer(_));
   EXPECT_CALL(*test_sessions_[0]->timeout_timer_, disableTimer());
+  EXPECT_CALL(*event_logger_, logDegraded(_, _));
```
Comment (junr03) [inline test/common/upstream/health_checker_impl_test.cc]: Should we inspect the logs like in: https://github.com/envoyproxy/envoy/blob/7c14fd1f1a12333712b3e8a279d75cba22a6345d/test/common/upstream/outlier_detection_impl_test.cc#L773

**Comment by snowp** [comment]
@junr03 Docs are here: https://github.com/envoyproxy/envoy/pull/5367/files

File: source/common/upstream/health_checker_base_impl.cc (hunk 4 above)
Comment (snowp) [reply source/common/upstream/health_checker_base_impl.cc]: I went a bit further and also DRYd up the json serialization and writing, let me know what you think

**Comment by snowp** [comment]
Seems like the github suggestions don't set up DCO properly :( force pushed to fix

File: source/common/upstream/health_checker_base_impl.cc [hunk 7]
```diff
@@ -132,6 +143,7 @@ void HealthCheckerImplBase::refreshHealthyStat() {
   // Each hot restarted process health checks independently. To make the stats easier to read,
   // we assume that both processes will converge and the last one that writes wins for the host.
   stats_.healthy_.set(local_process_healthy_);
+  stats_.degraded_.set(local_process_degraded_);
```
Comment (mattklein123) [inline source/common/upstream/health_checker_base_impl.cc]: cc @jmarantz I would take a look at this PR. This is a very succinct example of the potential differences in gauges if we get rid of shard memory on hot restart.

**Pull request merged**